File: 4sum-ii.java
// https://leetcode.com/problems/4sum-ii/description/
// 4sum = 2sum + 2sum
// Time: O(n^2), Space: O(n^2)
class Solution {
    public int fourSumCount(int[] A, int[] B, int[] C, int[] D) {
        int r = 0;
        HashMap<Integer,Integer> m = new HashMap<Integer,Integer>();
        for(int a : A) {
        	for(int b : B) {
        		int sum = a+b;
        		m.put(sum, m.getOrDefault(sum, 0)+1);
        	}
        }
        for(int c : C) {
        	for(int d : D) {
        		int sum = c+d;
        		r += m.getOrDefault(-sum,0);
        	}
        }
        return r;
    }
}

File: 4sum-ii.py
# https://leetcode.com/problems/4sum-ii/description/
# 4 sum = 2 sum + 2 sum
# dictionary {} : put O(1), get O(1)
# Time: O(n^2), Space: O(n^2)
class Solution:
    def fourSumCount(self, A, B, C, D):
        dic = {}
        for i in range(len(A)):
            for j in range(len(B)):
                sum = A[i]+B[j]
                if sum in dic: dic[sum] += 1
                else: dic[sum] = 1
        cnt = 0
        for i in range(len(C)):
            for j in range(len(D)):
                sum = -C[i]-D[j]
                if sum in dic: cnt += dic[sum]
        return cnt
